src/main/java/com/google/api/codegen/config/Resources.java

```diff
@@ -80,9 +82,27 @@
   public static String getEntityName(FieldSegment segment) {
     Preconditions.checkArgument(isTemplateFieldSegment(segment));
     // TODO(shinfan): Consider finding a better way to determine the name if possible.
-    List<String> params = getWildcardCollectionNames(segment);
-    String lastParam = params.get(params.size() - 1);
-    return Inflector.singularize(lastParam);
+    List<LiteralWildcardSegment> literalWildcardSegments =
+        getWildcardCollectionMapFieldSegment(segment);
+    return literalWildcardSegments.get(literalWildcardSegments.size() - 1).paramName();
+  }
+
+  public static Map<String, String> getResourceToEntityNameMap(Iterable<FieldSegment> segments) {
+    Set<String> usedNameSet = new HashSet<>();
+    Map<String, String> nameMap = new LinkedHashMap<>();
+    for (FieldSegment segment : segments) {
+      String resourceNameString = Resources.templatize(segment);
+      String entityNameStringBase = Resources.getEntityName(segment);
+      String entityNameString = entityNameStringBase;
+      int index = 1;
```

Comment:
Another alternative is instead of appending numbers when there is a clash, instead start adding earlier literals. In the library case, this would produce something like shelf_book and archive_book.